x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/LuceneSourceOperator.java

```diff
@@ -85,6 +99,76 @@ public String describe() {
                 + scoreMode
                 + "]";
         }
+
+        /**
+         * Pick a strategy for the {@link DataPartitioning#AUTO} partitioning.
+         */
+        public static Function<Query, LuceneSliceQueue.PartitioningStrategy> autoStrategy(int limit) {
+            return limit == NO_LIMIT ? Factory::highSpeedAutoStrategy : Factory::lowOverheadAutoStrategy;
+        }
+
+        /**
+         * Use the {@link LuceneSliceQueue.PartitioningStrategy#SHARD} strategy because
+         * it has the lowest overhead. Used when there is a {@code limit} on the operator
+         * because that's for cases like {@code FROM foo | LIMIT 10} or
+         * {@code FROM foo | WHERE a == 1 | LIMIT 10} when the {@code WHERE} can be pushed
+         * to Lucene. In those cases we're better off with the lowest overhead we can
+         * manage - and that's {@link LuceneSliceQueue.PartitioningStrategy#SHARD}.
+         */
+        private static LuceneSliceQueue.PartitioningStrategy lowOverheadAutoStrategy(Query query) {
+            return LuceneSliceQueue.PartitioningStrategy.SHARD;
+        }
+
+        /**
+         * Select the {@link LuceneSliceQueue.PartitioningStrategy} based on the {@link Query}.
+         * <ul>
+         *     <li>
+         *         If the {@linkplain Query} matches <strong>no</strong> documents then this will
+         *         use the {@link LuceneSliceQueue.PartitioningStrategy#SEGMENT} strategy so we
+         *         minimize the overhead of finding nothing.
+         *     </li>
+         *     <li>
+         *         If the {@linkplain Query} matches <strong>all</strong> documents then this will
+         *         use the {@link LuceneSliceQueue.PartitioningStrategy#DOC} strategy because the
+         *         overhead of using that strategy for {@link MatchAllDocsQuery} is very low, and
+         *         we need as many CPUs as we can get to process all the documents.
+         *     </li>
+         *     <li>
+         *         Otherwise use the {@link LuceneSliceQueue.PartitioningStrategy#SEGMENT} strategy
+         *         because it's overhead is generally low.
+         *     </li>
+         * </ul>
+         */
+        private static LuceneSliceQueue.PartitioningStrategy highSpeedAutoStrategy(Query query) {
+            if (query instanceof ConstantScoreQuery c) {
```

Comment:
NIT: this could be a switch with patterns, similar to https://github.com/elastic/elasticsearch/pull/120875/files